File: Maths_And_Stats/Calculus/Euler_Approximation/Euler_Approximation.cs
using System;

class Program
{
    /// <summary>
    /// Main application function.
    /// </summary>
    static void Main()
    {
        // approximates f(2) given dy/dx = x + y; y(-1) = 1
        //  using 6 equal subdivisions
        Console.WriteLine(Approximate(-1, 1, 2, 6));
    }
    /// <summary>
    /// Uses Euler's method to approximate a point.
    /// </summary>
    /// <param name="xInitial">initial condition for the X value</param>
    /// <param name="yInitial">initial condition for the Y value</param>
    /// <param name="xApprox">xApprox value to approximate</param>
    /// <param name="steps">steps number of steps to use for the approximation. Used to determine step size.</param>
    /// <returns>approximation of f(xApprox)</returns>
    public static double Approximate(double xInitial, double yInitial, double xApprox, int steps)
    {
        // find step size
        // (b - a) / n
        double stepSize = (xApprox - xInitial) / (double)steps;

        // check track of the X and Y coordinates
        double prevX = xInitial;
        double prevY = yInitial;

        // iterate until we meet our end coordinate
        for (int i = 0; i < steps; i++)
        {
            // apply Euler's method
            // f(x) = f(prevY) + stepSize * dy/dx(prevX, prevY)
            prevY += stepSize * Differential(prevX, prevY);
            prevX += stepSize;
        }

        return prevY;
    }
    /// <summary>
    /// Represents dy/dx. Change this function to your desired differential equation
    /// </summary>
    /// <param name="x">X coordinate value</param>
    /// <param name="y">Y coordinate value</param>
    /// <returns>dy/dx evaluated at the coordinate (x, y)</returns>
    private static double Differential(double x, double y)
    {
        return (x + y);
    }
}

File: Maths_And_Stats/Calculus/Euler_Approximation/Euler_Approximation.cpp
/* CPP  Program to find approximation 
   of a ordinary differential equation 
   using euler method.*/
#include <iostream> 
using namespace std; 
  
// Consider a differential equation 
// dy/dx=(x + y + xy) 
float func(float x, float y) 
{ 
    return (x + y + x * y); 
} 
  
// Function for Euler formula 
void euler(float x0, float y, float h, float x) 
{ 
    float temp = -0; 
  
    // Iterating till the point at which we 
    // need approximation 
    while (x0 < x) { 
        temp = y; 
        y = y + h * func(x0, y); 
        x0 = x0 + h; 
    } 
  
    // Printing approximation 
    cout << "Approximate solution at x = "
         << x << "  is  " << y << endl; 
} 
  
// Driver program 
int main() 
{ 
    // Initial Values 
    float x0; 
    float y0; 
    float h;
    cin>>x0>>y0>>h;
  
    // Value of x at which we need approximation 
    float x ;
    cin>>x;
  
    euler(x0, y0, h, x); 
    return 0; 
} 
